lib/reek/configuration/app_configuration.rb

```diff
@@ -10,135 +14,95 @@ module Configuration
     #
     # @api private
     class AppConfiguration
+      include ConfigurationValidator
       EXCLUDE_PATHS_KEY = 'exclude_paths'
-      attr_reader :exclude_paths, :default_directive, :directory_directives
+      private_attr_writer :directory_directives, :default_directive, :excluded_paths
 
-      # Given this configuration file:
+      # Instantiate a configuration via given path.
       #
-      # ---
-      # IrresponsibleModule:
-      #   enabled: false
-      # "app/helpers":
-      #   UtilityFunction:
-      #     enabled: false
-      # exclude_paths:
-      #   - "app/controllers"
+      # @param path [Pathname] the path to the config file
       #
-      # this would result in the following configuration:
-      #
-      # exclude_paths = [ Pathname('spec/samples/two_smelly_files') ]
-      # default_directive = { Reek::Smells::IrresponsibleModule => { "enabled" => false } }
-      # directory_directives = { Pathname("spec/samples/three_clean_files/") =>
-      #                          { Reek::Smells::UtilityFunction => { "enabled" => false } } }
+      # @return [AppConfiguration]
+      def self.from_path(path: nil)
+        allocate.tap do |instance|
+          instance.instance_eval { find_and_load(path: path) }
+        end
+      end
+
+      # Instantiate a configuration by passing everything in.
       #
+      # @param directory_directives [Hash] for instance:
+      #   { Pathname("spec/samples/three_clean_files/") =>
+      #     { Reek::Smells::UtilityFunction => { "enabled" => false } } }
+      # @param default_directive [Hash] for instance:
+      #   { Reek::Smells::IrresponsibleModule => { "enabled" => false } }
+      # @param excluded_paths [Array] for instance:
+      #   [ Pathname('spec/samples/two_smelly_files') ]
       #
-      # @param options [OpenStruct]
-      #   e.g.: #<OpenStruct report_format =: text,
-      #                      location_format =: numbers,
-      #                      colored = true,
-      #                      smells_to_detect = [],
-      #                      config_file = #<Pathname:config/defaults.reek>,
-      #                      argv = [ "lib/reek/spec" ]>
-      def initialize(options = nil)
-        @directory_directives = {}
-        @default_directive    = {}
-        @exclude_paths        = []
+      # @return [AppConfiguration]
+      def self.from_map(directory_directives: {},
+                        default_directive: {},
+                        excluded_paths: [])
+        allocate.tap do |instance|
+          instance.instance_eval do
+            self.directory_directives = directory_directives.extend(DirectoryDirectives)
+            self.default_directive    = default_directive.extend(DefaultDirective)
+            self.excluded_paths       = excluded_paths.extend(ExcludedPaths)
+          end
+        end
+      end
 
-        load options
+      def self.default
+        from_path path: nil
       end
 
+      def initialize(*)
```

Comment:
I think allocate is incredibly helpful for exactly this use case since ruby is single-dispatch and not double-dispatch (which I would prefer, but I'm not Matz unfortunately) : factory methods. Actually, I can't think of any other uses cases right now but for this one it really shines :laughing: 

I think this SO post sums it up pretty nicely: http://stackoverflow.com/questions/13059552/classallocate-and-its-uses

>> Despite what the documentation says, the existence of the allocate method is not so much about memory management as it is about providing some finer grained control over the object creation lifecycle. Most of the time, you won't need this feature, but it is useful for certain edge cases.